File: app/src/main/java/com/example/vladimir/easyenglishlearn/exercises/WordQuizViewModel.java
package com.example.vladimir.easyenglishlearn.exercises;

import android.arch.lifecycle.LiveData;
import android.databinding.ObservableField;

import com.example.vladimir.easyenglishlearn.SingleLiveEvent;
import com.example.vladimir.easyenglishlearn.model.Word;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import static com.example.vladimir.easyenglishlearn.Constants.ANSWERS_COUNT;

public class WordQuizViewModel extends ExerciseViewModel {

    private SingleLiveEvent<Void> mClearRadioGroupLiveData;
    public final ObservableField<List<String>> answers = new ObservableField<>();


    public WordQuizViewModel() {
        mClearRadioGroupLiveData = new SingleLiveEvent<>();
    }

    public void onAnswerChecked(String answer) {
        mAnswerBuilder.append(answer);
        checkAnswer();
    }

    @Override
    void prepareQuestionAndAnswers() {
        super.prepareQuestionAndAnswers();

        List<String> answerList = new ArrayList<>();
        addAnswer(answerList, mCurrentWord);
        while (answerList.size() < ANSWERS_COUNT) {
            int randomIndex = new Random().nextInt(mWordList.size());
            Word tempWord = mWordList.get(randomIndex);
            if (isWordNotInAnswerList(answerList, tempWord)) addAnswer(answerList, tempWord);
        }
        Collections.shuffle(answerList);
        answers.set(answerList);
    }

    @Override
    void cleanPreviousData() {
        super.cleanPreviousData();
        mClearRadioGroupLiveData.call();
    }

    private boolean isWordNotInAnswerList(List<String> answerList, Word word) {
        return !answerList.contains(mTranslationDirection
                ? word.getTranslation()
                : word.getLexeme());
    }

    private void addAnswer(List<String> answerList, Word word) {
        answerList.add(mTranslationDirection
                ? word.getTranslation()
                : word.getLexeme());
    }

    LiveData<Void> getClearRadioGroupLiveData() {
        return mClearRadioGroupLiveData;
    }
}

File: app/src/main/java/com/vladimir_x/easyenglishlearn/exercises/WordQuizViewModel.kt
package com.vladimir_x.easyenglishlearn.exercises

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.vladimir_x.easyenglishlearn.SingleLiveEvent
import com.vladimir_x.easyenglishlearn.model.Word

class WordQuizViewModel : ExerciseViewModel() {
    private val _clearRadioGroupLiveData: SingleLiveEvent<Unit> = SingleLiveEvent()
    val clearRadioGroupLiveData: LiveData<Unit?>
        get() = _clearRadioGroupLiveData

    private val _questionLiveData = MutableLiveData<Pair<String, List<String>>>()
    val questionLiveData: LiveData<Pair<String, List<String>>>
        get() = _questionLiveData

    private var answers: List<String> = emptyList()

    fun onAnswerChecked(answerIndex: Int) {
        answerBuilder.append(answers[answerIndex])
        checkAnswer()
    }

    override fun prepareQuestionAndAnswers() {
        super.prepareQuestionAndAnswers()
        val answerList = wordList
            .filter { it != currentWord }
            .take(2)
            .toMutableList()
        currentWord?.let { answerList.add(it) }
        answers = answerList.map { convertWordToAnswer(it) }.shuffled()

        _questionLiveData.value = Pair(
            convertWordToQuestion(currentWord),
            answers
        )
    }

    override fun cleanPreviousData() {
        super.cleanPreviousData()
        _clearRadioGroupLiveData.call()
    }

    private fun convertWordToAnswer(word: Word): String {
        return if (translationDirection) word.translation else word.lexeme
    }

    private fun convertWordToQuestion(word: Word?): String {
        return if (translationDirection) word?.lexeme ?: "" else word?.translation ?: ""
    }
}